Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -446,4 +451,716 @@ alias nonsingular_of_affine_of_Z_ne_zero := nonsingular_of_Z_ne_zero
+    pow_ne_zero 2 hPz
+
+lemma isUnit_dblU_of_Y_eq {P Q : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
+    (hx : P x * Q z = Q x * P z) (hy : P y * Q z = Q y * P z) (hy' : P y * Q z = W.negY Q * P z) :
+    IsUnit (W.dblU P) :=
+  (dblU_ne_zero_of_Y_eq hP hPz hQz hx hy hy').isUnit
+
+variable (W') in
+/-- The $Z$-coordinate of a representative of `2 • P` for a point `P`. -/
+def dblZ (P : Fin 3 → R) : R :=
+  P z * (P y - W'.negY P) ^ 3
+
+lemma dblZ_smul (P : Fin 3 → R) (u : R) : W'.dblZ (u • P) = u ^ 4 * W'.dblZ P := by
+  simp only [dblZ, negY_smul, smul_fin3_ext]
+  ring1
+
+lemma dblZ_of_Z_eq_zero {P : Fin 3 → R} (hPz : P z = 0) : W'.dblZ P = 0 := by
+  rw [dblZ, hPz, zero_mul]
+
+lemma dblZ_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z)
+    (hy : P y * Q z = Q y * P z) (hy' : P y * Q z = W'.negY Q * P z) : W'.dblZ P = 0 := by
+  rw [dblZ, Y_eq_negY_of_Y_eq hQz hx hy hy', sub_self, zero_pow three_ne_zero, mul_zero]
+
+lemma dblZ_ne_zero_of_Y_ne [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : W'.Equation P)
+    (hQ : W'.Equation Q) (hPz : P z ≠ 0) (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z)
+    (hy : P y * Q z ≠ Q y * P z) : W'.dblZ P ≠ 0 :=
+  mul_ne_zero hPz <| pow_ne_zero 3 <| sub_ne_zero.mpr <| Y_ne_negY_of_Y_ne hP hQ hPz hQz hx hy
+
+lemma isUnit_dblZ_of_Y_ne {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z) (hy : P y * Q z ≠ Q y * P z) : IsUnit (W.dblZ P) :=
+  (dblZ_ne_zero_of_Y_ne hP hQ hPz hQz hx hy).isUnit
+
+lemma dblZ_ne_zero_of_Y_ne' [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : W'.Equation P)
+    (hQ : W'.Equation Q) (hPz : P z ≠ 0) (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z)
+    (hy : P y * Q z ≠ W'.negY Q * P z) : W'.dblZ P ≠ 0 :=
+  mul_ne_zero hPz <| pow_ne_zero 3 <| sub_ne_zero.mpr <| Y_ne_negY_of_Y_ne' hP hQ hPz hQz hx hy
+
+lemma isUnit_dblZ_of_Y_ne' {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z) (hy : P y * Q z ≠ W.negY Q * P z) :
+    IsUnit (W.dblZ P) :=
+  (dblZ_ne_zero_of_Y_ne' hP hQ hPz hQz hx hy).isUnit
+
+private lemma toAffine_slope_of_eq {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q)
+    (hPz : P z ≠ 0) (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z) (hy : P y * Q z ≠ W.negY Q * P z) :
+    W.toAffine.slope (P x / P z) (Q x / Q z) (P y / P z) (Q y / Q z) =
+      -eval P W.polynomialX / P z / (P y - W.negY P) := by
+  have hPy : P y - W.negY P ≠ 0 := sub_ne_zero.mpr <| Y_ne_negY_of_Y_ne' hP hQ hPz hQz hx hy
+  simp only [X_eq_iff hPz hQz, ne_eq, Y_eq_iff' hPz hQz] at hx hy
+  rw [Affine.slope_of_Y_ne hx <| negY_of_Z_ne_zero hQz ▸ hy, ← negY_of_Z_ne_zero hPz]
+  field_simp [eval_polynomialX, hPz]
+  ring1
+
+variable (W') in
+/-- The $X$-coordinate of a representative of `2 • P` for a point `P`. -/
+noncomputable def dblX (P : Fin 3 → R) : R :=
```

Comment:
Yeah that was the original expression that I called `preDblX` (and `preAddX` for addition). I thought that since we're already having massive polynomials, might as well write them out in full. Is there a use for `preDblX` by itself?

Reply:
Well your `dblZ` is still in factored form. Why not factor `dblX` as well to make it shorter?

Reply:
Better: the cubic factor in `dblX` can be replaced by 
```
-a2 x^3 + a1 x^2 y + x y^2 - a1 a3 x^2 z - 3 a4 x^2 z + a3 x y z - 
 2 a3^2 x z^2 - 9 a6 x z^2 - a2 a3^2 z^3 + a1 a3 a4 z^3 + a4^2 z^3 - 
 a1^2 a6 z^3 - 4 a2 a6 z^3
```
which is
`x polynomialY^2 - ψh₃ - 3 (x/z) polynomial`.
Now trying to do the same for `negDblY` ...